src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/meta/ConstantPool.java

```diff
@@ -211,6 +228,14 @@ default BootstrapMethodInvocation lookupBootstrapMethodInvocation(int index, int
         throw new UnsupportedOperationException();
     }
 
+    /**
+     * Returns all BootstrapMethodInvocations within the constant pool where
+     * {@link BootstrapMethodMethodInvocation#isInvokeDynamic} is true. If this
+     * constant pool contains no invokedynamic BootstrapMethodInvocations, then null
+     * is returned.
+     */
+    BootstrapMethodInvocation[] lookupAllIndyBootstrapMethodInvocations();
```

Comment:
Why not make this return all `BootstrapMethodInvocation`s? The caller can then filter out the indy ones with `isInvokeDynamic`. Also, please return a `List<BootstrapMethodInvocation>` instead of an array - we should never return arrays from JVMCI (see #23159 as an example of addressing existing API). Lastly, return `List.of()` instead of null.

Reply:
Changed to return a list.

> Why not make this return all BootstrapMethodInvocations
1. Within HotSpot it very easy to pick off all indy BootstrapMethodInvocations via [the ConstantPoolCache](https://github.com/openjdk/jdk/blob/72a3022dc6a1521d8e3f08fe5d592f760fc462d2/src/hotspot/share/oops/cpCache.hpp#L74)
1. Each invokedynamic bytecode location has an unique BootstrapMethodInvocation instance, but they may share the same constant pool entry, so it's not trivial to find all BootstrapMethodInvocations. One would have to iterate both all method bytecodes and constant pool slots

Reply:
I prototyped the option `List<BootstrapMethodInvocation> lookupBootstrapMethodInvocations(boolean indy)` here: https://github.com/openjdk/jdk/compare/master...teshull:jdk:jvmci_bootstrap_alternative

As part of this I also prototyped generic BSM resolution / lookup logic

From the SVM perspective, retrieving condys via this new support isn't a big win. It's easy enough already to walk the ConstantPool. However, for symmetry purposes, it is reasonable to have (along with the resolve / lookup). What's your preference, this new version or the original?

Reply:
I like the symmetry of the new version. Also, I think you can simplify things by replacing use of `flatMap` [here](https://github.com/openjdk/jdk/compare/master...teshull:jdk:jvmci_bootstrap_alternative#diff-b782878562668748c5c59acc2e937f7c24de4529b8a74bd3a4eae83fa0e07846R679) with `filter`.